[RateLimiter] Fix wait duration for fixed window policy

jlekowski opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.3
| Bug fix?      | yes
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tickets       | Fix #... <!-- prefix each issue number with "Fix #", no need to create an issue if none exist, explain below instead -->
| License       | MIT
| Doc PR        | symfony/symfony-docs#... <!-- required for new features -->

When using `fixed_window` policy and calling `->wait()` after overconsuming, there might be no `sleep()` called. Reproducer below:

```php
<?php

use Symfony\Component\RateLimiter\RateLimiterFactory;
use Symfony\Component\RateLimiter\Storage\InMemoryStorage;

require_once './vendor/autoload.php';


$factory = new RateLimiterFactory([
    'id' => 'some_id',
    'policy' => 'fixed_window',
    'limit' => 10,
    'interval' => '2 seconds',
], new InMemoryStorage());

$limiter = $factory->create();

$limit = $limiter->consume(8);

$limit = $limiter->consume(4);
$start = microtime(true);
$limit->wait();
echo 'WAITED: ' . 1000 * (microtime(true) - $start);
// WAITED: 0.0030994415283203 - instantaneous instead of WAITED: 2000.3020763397 after the fix
```

jlekowski committed:
Fix wait duration for fixed window policy

Nyholm commented:
Thank you. 
It is indeed a bug. But Im not sure you have the correct solution. 

Isn't the correct behaviour to not accept that you are trying to overconsume?

jlekowski commented:
>Isn't the correct behaviour to not accept that you are trying to overconsume?

@Nyholm, thanks for looking at this PR. I think you can always call `->consume()`, but you can check something like `var_dump($limit->getRemainingTokens(), $limit->isAccepted());` afterwards to see there are still 2 tokens remaining and that it has not been accepted. 

Here's a code I used to compare different policies:
```php
foreach (['fixed_window', 'sliding_window', 'token_bucket'] as $policy) {
    echo "\nPOLICY: $policy\n";

    $factory = new RateLimiterFactory([
        'id' => 'db_write',
        'policy' => $policy,
        'limit' => 10,
        'interval' => '2 seconds',
        'rate' => ['amount' => 10,'interval' => '2 seconds'],
    ], new InMemoryStorage());

    $limiter = $factory->create();

    $limit = $limiter->consume(8);

    $limit = $limiter->consume(4);
    echo "REMAINING TOKENS: " . $limit->getRemainingTokens() . PHP_EOL;
    echo "IS ACCEPTED: " . (int)$limit->isAccepted() . PHP_EOL;
    $start = microtime(true);
    $limit->wait();
    echo 'WAITED: ' . 1000 * (microtime(true) - $start) . PHP_EOL;
    $limit = $limiter->consume(4);
    echo "REMAINING TOKENS: " . $limit->getRemainingTokens() . PHP_EOL;
    echo "IS ACCEPTED: " . (int)$limit->isAccepted() . PHP_EOL;
}
```
BTW, there might be another issue with `sliding_window` policy, but it's not for this ticket.

jlekowski commented:
@Nyholm, would you have any suggestions for this PR? Could this be merged, or requires a change?

Also, could you have a look at https://github.com/symfony/symfony/issues/42194, please?

Nyholm commented on src/Symfony/Component/RateLimiter/Policy/Window.php:
```diff
@@ -80,15 +80,4 @@ public function getAvailableTokens(float $now)
         return $this->maxSize - $this->hitCount;
     }
-
-    public function calculateTimeForTokens(int $tokens): int
```
Removing this method would break BC. We must revert this.

Nyholm commented on src/Symfony/Component/RateLimiter/Policy/FixedWindowLimiter.php:
```diff
@@ -70,8 +70,7 @@ public function reserve(int $tokens = 1, float $maxTime = null): Reservation
             } else {
-                $remainingTokens = $tokens - $availableTokens;
-                $waitDuration = $window->calculateTimeForTokens($remainingTokens);
+                $waitDuration = $this->interval;
```
It seams like the issue is that `calculateTimeForTokens()` is looking at the "remaining tokens" while it should look for the total number of tokens the user wants to reserve. 

Would the following fix also work for you?
```suggestion
                $waitDuration = $window->calculateTimeForTokens($tokens);
```

Nyholm reviewed (changes requested):
I finally got my head around this PR. 

What do you think of changing the solution slightly? Will it still work?

jlekowski commented:
@Nyholm, thanks for having a closer look into this.

Fixing the number of tokens to calculate time for works too.

I made the changes according to your suggestions. If that looks fine, I will:
* squash my commits
* rebase the branch from 5.3 and point the PR to 5.3 branch (as the milestone here was set)

The reason behind removing `calculateTimeForTokens()` was that the class is marked as `@internal` and in the current flow `$window->calculateTimeForTokens($tokens);` will always return the equivalent of `$this->interval;` (see my comment in https://github.com/symfony/symfony/pull/42168/commits/44807a638aed81e458e9d98ef83f494fa5da9a53#diff-26af5f9f51ed8bd11b4b02e6d02eef2872e174b3ba50a6f6aa9edb218400ac13).
That means that even if we keep `calculateTimeForTokens()` for BC, we can avoid calling it.

Nyholm reviewed (approved):
Im happy with this. 

Please squash the commits and rebase on 5.3. Make sure that none of the test failures are related to this PR. Then I would be happy if @wouterj saw this PR too. 

About removing the method. 
I didn't see the class was internal, if so, you are correct. Removing a method from an internal class is not a BC break.

jlekowski commented:
Thank you @Nyholm. I squashed and rebased my changes.

> About removing the method. I didn't see the class was internal, if so, you are correct. Removing a method from an internal class is not a BC break.

@wouterj, please have a look how you feel about this change. My original commit was here https://github.com/jlekowski/symfony/commit/44807a638aed81e458e9d98ef83f494fa5da9a53 with `Symfony\Component\RateLimiter\Policy\Window::calculateTimeForTokens()` being removed as it does not need to be used in `FixedWindowLimiter`. But it's your party, so I'm fine with either change :)

fabpot commented:
Thank you @jlekowski.
